src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -764,5 +1564,169 @@ private static bool ShouldRoundUp(bool lsbBit, bool roundBit, bool hasTailBits)
             // two is even (this is "round-to-even").
             return roundBit && (hasTailBits || lsbBit);
         }
+
+
+        /// <summary>
+        /// Daniel Lemire's Fast-float algorithm please refer to https://arxiv.org/abs/2101.11408
+        /// Ojective is to calculate m and p, adjusted mantissa and power of 2, based on the
+        /// following equality : (m x 2^p) =  (w x 10^q)
+        /// </summary>
+        /// <param name="q">decimal exponent</param>
+        /// <param name="w">decimal significant (mantissa)</param>
+        /// <param name="info">parameters for calculations for the value's type (double, float, half)</param>
+        /// <returns>Tuple : Exponent (power of 2) and adjusted mantissa </returns>
+        internal static (int Exponent, ulong Mantissa) ComputeFloat(long q, ulong w, FloatingPointInfo info)
+        {
+            int exponent;
+            ulong mantissa = 0;
+
+            if ((w == 0) || (q < info.MinDecimalExponent))
+            {
+                // result should be zero
+                return default;
+            }
+            if (q > info.MaxDecimalExponent)
+            {
+                // we want to get infinity:
+                exponent = info.InfinitePower;
+                mantissa = 0;
+                return (exponent, mantissa);
+            }
+            // At this point in time q is in [smallest_power_of_five, largest_power_of_five].
+
+            // We want the most significant bit of i to be 1. Shift if needed.
+            int lz = BitOperations.LeadingZeroCount(w);
+            w <<= lz;
+
+            // The required precision is mantissa_explicit_bits() + 3 because
+            // 1. We need the implicit bit
+            // 2. We need an extra bit for rounding purposes
+            // 3. We might lose a bit due to the "upperbit" routine (result too small, requiring a shift)
+
+            var product = ComputeProductApproximation(info.DenormalMantissaBits + 3, q, w);
+            if (product.low == 0xFFFFFFFFFFFFFFFF)
+            {
+                // could guard it further
+                // In some very rare cases, this could happen, in which case we might need a more accurate
+                // computation that what we can provide cheaply. This is very, very unlikely.
+                //
+                bool inside_safe_exponent = (q >= -27) && (q <= 55); // always good because 5**q <2**128 when q>=0,
```

Comment:
nit: we should at least use the "standard" .NET naming conventions here and make this `insideSafeExponent` and below it should be `upperBit`.

Reply:
done !